pkg/kubectl/apply/element.go

```diff
@@ -0,0 +1,375 @@
+}
+
+// GetFieldMergeKeys implements FieldMeta.GetFieldMergeKeys
+func (s FieldMetaImpl) GetFieldMergeKeys() MergeKeys {
+	return s.MergeKeys
+}
+
+// GetFieldType implements FieldMeta.GetFieldType
+func (s FieldMetaImpl) GetFieldType() string {
+	return s.Type
+}
+
+// MergeKeyValue records the value of the mergekey for an item in a list
+type MergeKeyValue map[string]string
+
+// Equal returns true if the MergeKeyValues share the same value,
+// representing the same item in a list
+func (v MergeKeyValue) Equal(o MergeKeyValue) bool {
+	if len(v) != len(o) {
+		return false
+	}
+
+	for key, v1 := range v {
+		if v2, found := o[key]; !found || v1 != v2 {
+			return false
+		}
+	}
+
+	return true
+}
+
+// MergeKeys is the set of fields on an object that uniquely identify
+// and is used when merging lists to identify the "same" object
+// independent of the ordering of the objects
+type MergeKeys []string
+
+// GetMergeKeyValue parses the MergeKeyValue from an item in a list
+func (mk MergeKeys) GetMergeKeyValue(i interface{}) (MergeKeyValue, error) {
+	result := MergeKeyValue{}
+	if len(mk) <= 0 {
+		return result, fmt.Errorf("merge key must have at least 1 value to merge")
+	}
+	m, ok := i.(map[string]interface{})
+	if !ok {
+		return result, fmt.Errorf("cannot use mergekey %v for primitive item in list %v", mk, i)
+	}
+	for _, field := range mk {
+		if value, found := m[field]; !found {
+			result[field] = ""
+		} else {
+			result[field] = fmt.Sprintf("%v", value)
+		}
+	}
+	return result, nil
+}
+
+type source int
+
+const (
+	recorded source = iota
+	local
+	remote
+)
+
+// CombinedPrimitiveSlice implements a slice of primitives
+type CombinedPrimitiveSlice struct {
+	Items []*PrimitiveListItem
+}
+
+// PrimitiveListItem represents a single value in a slice of primitives
+type PrimitiveListItem struct {
+	// Value is the value of the primitive, should match Recorded, Local and Remote
+	Value interface{}
+
+	RawElementData
+}
+
+// Contains returns true if the slice contains the l
+func (s *CombinedPrimitiveSlice) lookup(l interface{}) *PrimitiveListItem {
+	val := fmt.Sprintf("%v", l)
+	for _, i := range s.Items {
+		if fmt.Sprintf("%v", i.Value) == val {
+			return i
+		}
+	}
+	return nil
+}
+
+func (s *CombinedPrimitiveSlice) upsert(l interface{}, source source) {
+	item := s.lookup(l)
+
+	// Append an item to the list if we didn't find it
+	if item == nil {
+		item = &PrimitiveListItem{
+			Value: l,
+		}
+		s.Items = append(s.Items, item)
+	}
+
+	// Update the appropriate value
+	switch source {
+	case recorded:
+		item.Recorded = l
+	case local:
+		item.Local = l
+	case remote:
+		item.Remote = l
+	}
+}
+
+// UpsertRecorded adds l to the slice.  If there is already a value of l in the
+// slice for either the local or remote, set on that value as the recorded value
+// Otherwise append a new item to the list with the recorded value.
+func (s *CombinedPrimitiveSlice) UpsertRecorded(l interface{}) {
+	s.upsert(l, recorded)
```

Comment:
trying to avoid the switch above (hehe). Absolutely not sure it would be better:
```
item := s.upsert(l)
item.Local = l
```
It *maybe* removes cyclomatic/cognitive complexity and the extra constants ... ??

Reply:
yeah that is better.  Will update.